examples/nlp/multi_label_classification.py

```diff
@@ -0,0 +1,382 @@
+particular entry.
+"""
+
+print(f"There are {len(arxiv_data)} rows in the dataset.")
+
+"""
+Real-world data is noisy. One of the most commonly observed such noise is data
+duplication. Here we notice that our initial dataset has got about 13k duplicate entries.
+
+"""
+
+total_duplicate_titles = sum(arxiv_data["titles"].duplicated())
+print(f"There are {total_duplicate_titles} duplicate titles.")
+
+"""
+Before proceeding further we first drop these entries. 
+"""
+
+arxiv_data = arxiv_data[~arxiv_data["titles"].duplicated()]
+print(f"There are {len(arxiv_data)} rows in the deduplicated dataset.")
+
+# There are some terms with occurrence as low as 1.
+print(sum(arxiv_data["terms"].value_counts() == 1))
+
+# How many unique terms?
+print(arxiv_data["terms"].nunique())
+
+"""
+As observed above, out of 3157 unique combinations of `terms`, 2321 entries have the
+lowest occurrence. To prepare our train, validation, and test sets with
+[stratification](https://en.wikipedia.org/wiki/Stratified_sampling), we need to drop
+these terms. 
+"""
+
+# Filtering the rare terms.
+arxiv_data_filtered = arxiv_data.groupby("terms").filter(lambda x: len(x) > 1)
+arxiv_data_filtered.shape
+
+"""
+## Convert the string labels to list of strings
+
+The initial labels are represented as raw strings. Here we make them `List[str]` for a
+more compact representation.
+"""
+
+arxiv_data_filtered["terms"] = arxiv_data_filtered["terms"].apply(
+    lambda x: literal_eval(x)
+)
+arxiv_data_filtered["terms"].values[:5]
+
+"""
+## Stratified splits because of class imbalance
+
+The dataset has a
+[class imbalance problem](https://developers.google.com/machine-learning/glossary/#class-imbalanced-dataset).
+So, to have a fair evaluation result, we need to ensure the datasets are sampled with
+stratification. To know more about different strategies to deal with the class imbalance
+problem, you can follow
+[this tutorial](https://www.tensorflow.org/tutorials/structured_data/imbalanced_data). 
+For an end-to-end demonstration of classification with imbablanced data, refer to 
+[Imbalanced classification: credit card fraud detection](https://keras.io/examples/structured_data/imbalanced_classification/).
+"""
+
+test_split = 0.1
+
+# Initial train and test split.
+train_df, test_df = train_test_split(
+    arxiv_data_filtered,
+    test_size=test_split,
+    stratify=arxiv_data_filtered["terms"].values,
+)
+
+# Splitting the test set further into validation
+# and new test sets.
+val_df = test_df.sample(frac=0.5)
+test_df.drop(val_df.index, inplace=True)
+
+print(f"Number of rows in training set: {len(train_df)}")
+print(f"Number of rows in validation set: {len(val_df)}")
+print(f"Number of rows in test set: {len(test_df)}")
+
+"""
+## Multi-label binarization
+
+Now we preprocess our labels using
+[`MultiLabelBinarizer`](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MultiLabelBinarizer.html). 
+"""
+
+mlb = MultiLabelBinarizer()
```

Comment:
Sure. So, here's what we tried:

```python
max_seqlen = 150
batch_size = 128
padding_token = "<pad>"
auto = tf.data.AUTOTUNE


def unify_text_length(text):
    # Split the given abstract and calculate its length.
    word_splits = tf.strings.split(text, sep=" ")
    sequence_length = tf.shape(word_splits)[0]
    
    # Calculate the padding amount.
    padding_amount = max_seqlen - sequence_length
    
    # Check if we need to pad or truncate.
    if padding_amount > 0:
        unified_text = tf.pad([text], [[0, padding_amount]], constant_values="<pad>")
        unified_text = tf.strings.reduce_join(unified_text, separator="")
    else:
        unified_text = tf.strings.reduce_join(word_splits[:max_seqlen], separator=" ")
    
    # The expansion is needed for subsequent vectorization.
    return tf.expand_dims(unified_text, -1)


def preprocess_label(label):
    return binarizer([label])


def preprocess_text_and_label(text, label):
    unified_text = unify_text_length(text)
    preprocessed_label = preprocess_label(label)


def make_dataset(dataframe, is_train=True):
    dataset = tf.data.Dataset.from_tensor_slices(
        (dataframe["summaries"].values, dataframe["terms"].values)
    )
    dataset = dataset.shuffle(batch_size * 10) if is_train else dataset
    dataset = dataset.map(preprocess_text_and_label, num_parallel_calls=auto).cache()
    return dataset.batch(batch_size).prefetch(auto)
```

where the `binarizer` is defined as follows:

```python
train_term_set = set()
for terms in dataframe["terms"]:
    for i in range(len(terms)):
        train_term_set.add(terms[i])

train_term_set = tf.convert_to_tensor(np.array(list(train_term_set)))

binarizer = layers.StringLookup(output_mode="multi_hot")
binarizer.adapt(train_term_set)
```

Now when we call `train_dataset = make_dataset(train_df, is_train=True)` we are met with:

```
ValueError: Failed to convert a NumPy array to a Tensor (Unsupported object type list).
```

This is why we opted for `MultiLabelBinarizer`. Happy to consider any better method you may suggest.

Reply:
Here's another approach we tried and it works:

```python
max_seqlen = 150
batch_size = 128
padding_token = "<pad>"
auto = tf.data.AUTOTUNE


def unify_text_length(text, label):
    # Split the given abstract and calculate its length.
    word_splits = tf.strings.split(text, sep=" ")
    sequence_length = tf.shape(word_splits)[0]
    
    # Calculate the padding amount.
    padding_amount = max_seqlen - sequence_length
    
    # Check if we need to pad or truncate.
    if padding_amount > 0:
        unified_text = tf.pad([text], [[0, padding_amount]], constant_values="<pad>")
        unified_text = tf.strings.reduce_join(unified_text, separator="")
    else:
        unified_text = tf.strings.reduce_join(word_splits[:max_seqlen], separator=" ")
    
    # The expansion is needed for subsequent vectorization.
    return tf.expand_dims(unified_text, -1), label


def preprocess_label(labels):
    preprocessed_labels = []
    for label in labels:
        preprocessed_label = binarizer([label]).numpy()
        preprocessed_labels.append(preprocessed_label)
    return np.array(preprocessed_labels)


def make_dataset(dataframe, is_train=True):
    preprocessed_labels = preprocess_label(dataframe["terms"].values)
    dataset = tf.data.Dataset.from_tensor_slices(
        (dataframe["summaries"].values, preprocessed_labels)
    )
    dataset = dataset.shuffle(batch_size * 10) if is_train else dataset
    dataset = dataset.map(unify_text_length, num_parallel_calls=auto).cache()
    return dataset.batch(batch_size).prefetch(auto)
```

But now, we are having to iterate through the entire dataset first to process the labels with the `binarizer` and then do the rest.

Reply:
@fchollet a gentle ping for the above comment for your suggestion.

Reply:
The crux of the issue is converting the variable-length pandas column to a Tensor. @mattdangerw created this colab with a fix: https://colab.sandbox.google.com/drive/1VfqrPxm-aFBy6nqzFWSklqUW7naA7DpQ?resourcekey=0-FTknJ6xd6qhbstVVD0iaBA#scrollTo=1vgxbdwGf07E

The key part is this: you need to cast the variable-length pandas column to a ragged tensor.

```python
terms = tf.ragged.constant(train_df["terms"].values)
```

Since you also need to convert a multi-hot encoded output back to a list of labels: you'll need to write your own function for this, e.g.

```python
terms = tf.ragged.constant(train_df["terms"].values)
lookup = tf.keras.layers.StringLookup(output_mode='multi_hot')
lookup.adapt(terms)
vocab = lookup.get_vocabulary()

def invert_multi_hot(encoded_labels):
  """Reverse a single multi-hot encoded label to a tuple of vocab terms."""
  hot_indices = np.argwhere(encoded_labels == 1.0)[..., 0]
  return np.take(vocab, hot_indices)
```

Reply:
@fchollet @mattdangerw could you provide access to the Colab mentioned in the comment above?

Reply:
Hey! Sorry yeah not sure if I can share that version, but try this link.
https://colab.research.google.com/gist/mattdangerw/c2086b110a94918df1362f74697efba2/multi_label_trainer_tfidf-with-stringlookup.ipynb

Reply:
@mattdangerw a small nit was to actually use the `lookup` while creating the datasets (referring to the `make_dataset()` utility):

```
label_binarized = lookup(labels).numpy()
```

in place of 

```
label_binarized = mlb(labels)
```

But on the other hand, if you'd like to push your changes to the PR please do so and we will be more than happy to have you as a co-author. In fact, I think that should be the case since your changes and contributions to the example are quite non-trivial in my eyes.

Reply:
Oops, yeah must have not saved with actually using the lookup there.

Just took a look at the code changes looks good to me, one nit I found. Thanks very much for updating and thanks for the acknowledgement on the end! No need for authorship, this was a small change to a big example.

Thanks!